--- a/squashfs-tools/squashfs_fs.h
+++ b/squashfs-tools/squashfs_fs.h
@@ -26,16 +26,16 @@
 #define SQUASHFS_CACHED_FRAGMENTS	CONFIG_SQUASHFS_FRAGMENT_CACHE_SIZE	
 #define SQUASHFS_MAJOR			4
 #define SQUASHFS_MINOR			0
-#define SQUASHFS_MAGIC			0x73717368
-#define SQUASHFS_MAGIC_SWAP		0x68737173
+#define SQUASHFS_MAGIC			0x74696e69
+#define SQUASHFS_MAGIC_SWAP		0x696e6974
 #define SQUASHFS_START			0
 
 /*
  * Squashfs + LZMA
  */
 
-#define SQUASHFS_MAGIC_LZMA		0x71736873
-#define SQUASHFS_MAGIC_LZMA_SWAP	0x73687371
+#define SQUASHFS_MAGIC_LZMA		0x6974696e
+#define SQUASHFS_MAGIC_LZMA_SWAP	0x6e697469
 
 /* size of metadata (inode and directory) blocks */
 #define SQUASHFS_METADATA_SIZE		8192
